pkg/kudoctl/cmd/diagnostics/print.go

```diff
@@ -0,0 +1,195 @@
+package diagnostics
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"strings"
+
+	"github.com/kudobuilder/kudo/pkg/kudoctl/util/kudo"
+
+	"github.com/spf13/afero"
+	"gopkg.in/yaml.v2"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/cli-runtime/pkg/printers"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+const (
+	DiagDir = "diag"
+	KudoDir = "diag/kudo"
+)
+
+type printMode int
+
+const (
+	ObjectWithDir      printMode = iota // print object into its own nested directory based on its name and kind
+	ObjectListWithDirs                  // print each object into its own nested directory based on its name and kind
+	RuntimeObject                       // print as a file based on its kind only
+)
+
+// nonFailingPrinter - print provided data into provided directory and accumulate errors instead of returning them.
+// Creates a nested directory if an object type requires so.
+type nonFailingPrinter struct {
+	fs     afero.Fs
+	errors []string
+}
+
+func (p *nonFailingPrinter) printObject(o runtime.Object, parentDir string, mode printMode) {
+	switch mode {
+	case ObjectWithDir:
+		if err := printSingleObject(p.fs, o, parentDir); err != nil {
+			p.errors = append(p.errors, err.Error())
+		}
+	case ObjectListWithDirs:
+		err := meta.EachListItem(o, func(ro runtime.Object) error {
+			if err := printSingleObject(p.fs, ro, parentDir); err != nil {
+				p.errors = append(p.errors, err.Error())
+			}
+			return nil
+		})
+		if err != nil {
+			p.errors = append(p.errors, err.Error())
+		}
+	case RuntimeObject:
+		fallthrough
+	default:
+		if err := printSingleRuntimeObject(p.fs, o, parentDir); err != nil {
+			p.errors = append(p.errors, err.Error())
+		}
+	}
+}
+
+func (p *nonFailingPrinter) printError(err error, parentDir, name string) {
+	b := []byte(err.Error())
+	if err := printBytes(p.fs, b, parentDir, fmt.Sprintf("%s.err", name)); err != nil {
+		p.errors = append(p.errors, err.Error())
+	}
+}
+
+func (p *nonFailingPrinter) printLog(log io.ReadCloser, parentDir, name string) {
+	if err := printLog(p.fs, log, parentDir, name); err != nil {
+		p.errors = append(p.errors, err.Error())
+	}
+}
+
+func (p *nonFailingPrinter) printYaml(v interface{}, parentDir, name string) {
+	if err := printYaml(p.fs, v, parentDir, name); err != nil {
+		p.errors = append(p.errors, err.Error())
+	}
+}
+
+// printSingleObject - print a runtime.object assuming it exposes metadata by implementing metav1.object
+// or panic otherwise. object is put into a nested directory.
+func printSingleObject(fs afero.Fs, obj runtime.Object, parentDir string) error {
+	if !isKudoCR(obj) {
+		err := kudo.SetGVKFromScheme(obj, scheme.Scheme)
+		if err != nil {
+			return err
+		}
+	}
+
+	o, _ := obj.(object)
+	relToParentDir := fmt.Sprintf("%s_%s", strings.ToLower(o.GetObjectKind().GroupVersionKind().Kind), o.GetName())
```

Comment:
```suggestion
	o, _ := obj.(metav1.Object)
	relToParentDir := fmt.Sprintf("%s_%s", strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind), o.GetName())
```

Let's get rid of the `object` interface that merges `metav1.Object` and `runtime.Object`. It's really confusing, and it's only used to call `o.GetName()`.